File: static/solutions/insert_list.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <memory>
#include <vector>

#include "./Linked_list_prototype.h"

using std::make_shared;
using std::shared_ptr;
using std::vector;

// @include
// Insert new_node after node.
void InsertAfter(const shared_ptr<ListNode<int>>& node,
                 const shared_ptr<ListNode<int>>& new_node) {
  new_node->next = node->next;
  node->next = new_node;
}
// @exclude

void CheckAnswer(shared_ptr<ListNode<int>> L, const vector<int>& vals) {
  for (int i = 0; i < vals.size(); ++i) {
    assert(L->data == vals[i]);
    L = L->next;
  }
  assert(!L);
}

int main(int argc, char* argv[]) {
  shared_ptr<ListNode<int>> L;
  L = make_shared<ListNode<int>>(ListNode<int>{
      2, make_shared<ListNode<int>>(ListNode<int>{
             4, make_shared<ListNode<int>>(ListNode<int>{3, nullptr})})});
  InsertAfter(L, make_shared<ListNode<int>>(ListNode<int>{1, nullptr}));
  CheckAnswer(L, {2, 1, 4, 3});
  InsertAfter(L->next->next,
              make_shared<ListNode<int>>(ListNode<int>{10, nullptr}));
  CheckAnswer(L, {2, 1, 4, 10, 3});
  InsertAfter(L->next->next->next->next,
              make_shared<ListNode<int>>(ListNode<int>{-1, nullptr}));
  CheckAnswer(L, {2, 1, 4, 10, 3, -1});
  return 0;
}

File: static/python/insert_list.py
from linked_list_prototype import ListNode


# @include
# Insert new_node after node.
def insert_after(node, new_node):
    new_node.next = node.next
    node.next = new_node
# @exclude


def check_answer(L, vals):
    for val in vals:
        assert L.data == val
        L = L.next
    assert not L


def main():
    L = ListNode(2, ListNode(4, ListNode(3, None)))
    insert_after(L, ListNode(1, None))
    check_answer(L, [2, 1, 4, 3])
    insert_after(L.next.next, ListNode(10, None))
    check_answer(L, [2, 1, 4, 10, 3])
    insert_after(L.next.next.next.next, ListNode(-1, None))
    check_answer(L, [2, 1, 4, 10, 3, -1])


if __name__ == '__main__':
    main()
